rotator/rotator.go

```diff
@@ -0,0 +1,223 @@
+
+		logger.Infof("Checking that all %d nodes are running...", masterASG.DesiredCapacity)
+		err = FinalCheck(&masterASG, clientset, logger)
+		if err != nil {
+			return rotatorMetadata, err
+		}
+
+		logger.Infof("ASG %s rotated successfully.", masterASG.Name)
+	}
+
+	for index, workerASG := range rotatorMetadata.WorkerGroups {
+		logger.Infof("The autoscaling group %s has %d instance(s)", workerASG.Name, workerASG.DesiredCapacity)
+
+		err = WorkerNodeRotation(cluster, &workerASG, clientset, logger)
+		if err != nil {
+			rotatorMetadata.WorkerGroups[index] = workerASG
+			return rotatorMetadata, err
+		}
+
+		rotatorMetadata.WorkerGroups[index] = workerASG
+
+		logger.Infof("Checking that all %d nodes are running...", workerASG.DesiredCapacity)
+		err = FinalCheck(&workerASG, clientset, logger)
+		if err != nil {
+			return rotatorMetadata, err
+		}
+
+		logger.Infof("ASG %s rotated successfully.", workerASG.Name)
+	}
+
+	logger.Info("All ASGs rotated successfully")
+	return rotatorMetadata, nil
+}
+
+// FinalCheck checks that rotation is complete.
+func FinalCheck(autoscalingGroup *AutoscalingGroup, clientset *kubernetes.Clientset, logger *logrus.Entry) error {
+	asg, err := awsTools.AutoScalingGroupReady(autoscalingGroup.Name, autoscalingGroup.DesiredCapacity, logger)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get AutoscalingGroup ready")
+	}
+
+	asgNodes, err := awsTools.GetNodeHostnames(asg.Instances)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get node hostnames")
+	}
+
+	err = k8sTools.NodesReady(asgNodes, clientset, logger)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get cluster nodes ready")
+	}
+
+	return nil
+}
+
+// MasterNodeRotation handles rotation of master nodes.
+func MasterNodeRotation(cluster *model.Cluster, autoscalingGroup *AutoscalingGroup, clientset *kubernetes.Clientset, logger *logrus.Entry) error {
+
+	for len(autoscalingGroup.Nodes) > 0 {
+		logger.Infof("The number of nodes in the ASG to be rotated is %d", len(autoscalingGroup.Nodes))
+
+		nodesToRotate := []string{autoscalingGroup.Nodes[0]}
+
+		err := autoscalingGroup.DrainNodes(nodesToRotate, 10, int(cluster.EvictGracePeriod), cluster.WaitBetweenDrains, clientset, logger, "master")
+		if err != nil {
+			return err
+		}
+
+		err = awsTools.DetachNodes(false, nodesToRotate, autoscalingGroup.Name, logger)
+		if err != nil {
+			return err
+		}
+
+		err = awsTools.TerminateNodes(nodesToRotate, logger)
+		if err != nil {
+			return err
+		}
+
+		logger.Info("Sleeping 60 seconds for autoscaling group to balance...")
+		time.Sleep(60 * time.Second)
+
+		autoscalingGroupReady, err := awsTools.AutoScalingGroupReady(autoscalingGroup.Name, autoscalingGroup.DesiredCapacity, logger)
+		if err != nil {
+			return err
+		}
+
+		nodeHostnames, err := awsTools.GetNodeHostnames(autoscalingGroupReady.Instances)
+		if err != nil {
+			return err
+		}
+
+		newNodes := newNodes(nodeHostnames, autoscalingGroup.Nodes)
```

Comment:
Could you explain a bit the logic here? Why autoscaling.Nodes are the old nodes?

Reply:
@stafot sure, autoscaling.Nodes is the only source of truth that we need. It includes a list of all the nodes that the ASG included when we decided to initiate the rotation. This is the list against which we compare and check the progress of rotation.